tinygrad/graph.py

```diff
@@ -1,76 +1,72 @@
 import os
 import atexit
 import itertools
+import networkx as nx  # type: ignore
 from collections import defaultdict
 from typing import Dict, List
-from tinygrad.ops import DeviceBuffer, DEBUG, UnaryOps, BinaryOps, ReduceOps, MovementOps, ProcessingOps, LoadOps, Op, OpType, LazyOp, get_buffers, get_lazyops
+from tinygrad.ops import DeviceBuffer, DEBUG, GlobalCounters, UnaryOps, BinaryOps, ReduceOps, MovementOps, ProcessingOps, LoadOps, Op, OpType, LazyOp, get_buffers, get_lazyops
 from tinygrad.helpers import getenv
 
-GRAPH = getenv("GRAPH", 0)
+GRAPH, PRUNEGRAPH, GRAPHPATH = getenv("GRAPH", 0), getenv("PRUNEGRAPH", 0), getenv("GRAPHPATH", "/tmp/net")
 
 # **** debugging and graphing ****
 
+G = nx.DiGraph()
 cnts : Dict[OpType, int] = defaultdict(int)
 if GRAPH:
-  import networkx as nx  # type: ignore
-  G = nx.DiGraph()
   def save_graph_exit():
     for k,v in cnts.items():
       print(k, v)
-    if getenv("PRUNEGRAPH", 0):
-      dead_nodes = []
-      for n in G.nodes:
-        # prune movementops and loadops
-        if 'fillcolor' in G.nodes[n] and G.nodes[n]['fillcolor'] in ["#80ff8080", "#80ff80", "#FFFF8080", "#FFFF80"]:
-          for (x,_),(_,y) in itertools.product(G.in_edges(n), G.out_edges(n)):
-            G.add_edge(x, y)
-          dead_nodes.append(n)
-      for n in dead_nodes:
-        G.remove_node(n)
+    if PRUNEGRAPH:
+      prune_graph()
     print("saving", G)
-    nx.drawing.nx_pydot.write_dot(G, '/tmp/net.dot')
+    nx.drawing.nx_pydot.write_dot(G, f'{GRAPHPATH}.dot')
     # -Gnslimit=100 can make it finish, but you won't like results
-    os.system('dot -Tsvg /tmp/net.dot -o /tmp/net.svg')
+    os.system(f'dot -Tsvg {GRAPHPATH}.dot -o {GRAPHPATH}.svg')
   atexit.register(save_graph_exit)
 
-global_num_max = 0
 def nm(x):
-  global global_num_max
-  if not hasattr(x, 'global_num'):
-    setattr(x, 'global_num', global_num_max)
-    global_num_max += 1
-  return f"<{x.global_num}>"
+  if not hasattr(x, 'node_id'):
```

Comment:
Err, I wouldn't put this in GlobalCounters. There's no use for this outside the graph, and I don't think reset does something anyone wants. Rest looks good, revert that change and I'll merge

Reply:
Done